File: src/server/trash/provider/trv.ts
import type {
  Address,
  TrashCategory,
  TrashProvider,
  TrashScheduleEntry,
} from "../provider";

interface TRVAddress {
  id: string;
  adresse: string;
}

export interface Schedule {
  name: string;
  type: null;
  plan: null;
  calendar: Calendar[];
  deviations: null;
  plans_by_year: null;
  id: string;
  name_short: string;
}

export interface Calendar {
  dato: Date;
  frekvensType: number;
  frekvensIntervall: number;
  fraksjon: Fraksjon;
  fraksjonId: string;
  symbolId: number;
}

export enum Fraksjon {
  GlassOgMetallemballasje = "Glass- og metallemballasje",
  Matavfall = "Matavfall",
  PappOgPapir = "Papp og papir",
  Plastemballasje = "Plastemballasje",
  Restavfall = "Restavfall",
}

const fraksjonMap: Record<string, TrashCategory> = {
  [Fraksjon.GlassOgMetallemballasje]: "glass-metal",
  [Fraksjon.Matavfall]: "food",
  [Fraksjon.PappOgPapir]: "paper",
  [Fraksjon.Plastemballasje]: "plastic",
  [Fraksjon.Restavfall]: "rest",
};

export default class TRV implements TrashProvider {
  slug = "trv";
  name = "Trondheim Renholdsverk";
  logoUrl = "/trash-provider/trv.png";

  async getAddressId(address: Address): Promise<string> {
    const interpolated = `${address.streetName} ${address.houseNumber}`;

    const response = await fetch(
      `https://trv.no/wp-json/wasteplan/v2/adress/?s=${interpolated}`,
    );

    if (response.status !== 200) {
      throw new Error("Failed to fetch address");
    }

    const data = (await response.json()) as TRVAddress[];

    if (data.length === 0) {
      throw new Error("Address not found");
    }

    return data[0]!.id;
  }

  async getTrashSchedule(
    addressId: string,
    address: Address,
  ): Promise<TrashScheduleEntry[]> {
    const response = await fetch(
      `https://trv.no/wp-json/wasteplan/v2/calendar/${addressId}`,
    );

    if (!response.ok) {
      throw new Error("Failed to fetch trash schedule");
    }

    const data = (await response.json()) as Schedule;

    return data.calendar.map((entry) => ({
      date: entry.dato,
      type: fraksjonMap[entry.fraksjon]!,
    }));
  }
}

File: frontend/lib/repository/garbage_disposal/trv.dart
import 'package:dio/dio.dart';
import 'repository.dart';

final class TrvGarbageDisposalRepository implements GarbageDisposalRepository {
  final Dio _dio = Dio();

  TrvGarbageDisposalRepository();

  @override
  Future<String> getAddressId({required Address address}) async {
    try {
      final interpolated = '${address.streetName} ${address.streetNumber}';

      final response = await _dio.get(
        'https://trv.no/wp-json/wasteplan/v2/adress/',
        queryParameters: {'s': interpolated},
      );

      if (response.data == null) {
        throw const GarbageDisposalUnknownException(
          code: 'no_data',
          message: 'No data returned from server',
        );
      }

      final List<dynamic> addresses = response.data as List<dynamic>;

      if (addresses.isEmpty) {
        throw const GarbageDisposalNotFoundException(
          code: 'address_not_found',
          message: 'Address not found',
        );
      }

      final firstAddress = addresses.first as Map<String, dynamic>;
      return firstAddress['id'] as String;
    } on DioException catch (e) {
      throw _handleError(e);
    }
  }

  @override
  Future<List<TrashScheduleEntry>> getTrashSchedule({
    required String addressId,
  }) async {
    try {
      final response = await _dio.get(
        'https://trv.no/wp-json/wasteplan/v2/calendar/$addressId',
      );

      if (response.data == null) {
        throw const GarbageDisposalUnknownException(
          code: 'no_data',
          message: 'No data returned from server',
        );
      }

      final scheduleData = response.data as Map<String, dynamic>;
      final List<dynamic> calendar = scheduleData['calendar'] as List<dynamic>;

      return calendar.map((entry) {
        final entryMap = entry as Map<String, dynamic>;
        return TrashScheduleEntry(
          date: DateTime.parse(entryMap['dato'] as String),
          type: _mapFraksjonToCategory(entryMap['fraksjon'] as String),
        );
      }).toList();
    } on DioException catch (e) {
      throw _handleError(e);
    }
  }

  TrashCategory _mapFraksjonToCategory(String fraksjon) {
    switch (fraksjon) {
      case 'Glass- og metallemballasje':
        return TrashCategory.glassMetal;
      case 'Matavfall':
        return TrashCategory.food;
      case 'Papp og papir':
        return TrashCategory.cardboard;
      case 'Plastemballasje':
        return TrashCategory.plastic;
      case 'Restavfall':
        return TrashCategory.rest;
      default:
        // Default to rest for unknown categories
        return TrashCategory.rest;
    }
  }

  GarbageDisposalException _handleError(DioException e) {
    final response = e.response;

    if (response == null) {
      return GarbageDisposalUnknownException(
        code: 'network_error',
        message: e.message ?? 'Network error occurred',
      );
    }

    final data = response.data;

    // Try to extract error code and message from response
    String code = 'unknown';
    String message = 'An unknown error occurred';

    if (data is Map<String, dynamic>) {
      code = data['code']?.toString() ?? code;
      message = data['message']?.toString() ?? message;
    } else if (data is String) {
      message = data;
    }

    switch (response.statusCode) {
      case 400:
        return GarbageDisposalBadRequestException(code: code, message: message);
      case 404:
        return GarbageDisposalNotFoundException(code: code, message: message);
      case 500:
      case 502:
      case 503:
        return GarbageDisposalInternalServerException(
          code: code,
          message: message,
        );
      default:
        return GarbageDisposalUnknownException(code: code, message: message);
    }
  }
}
